### PR TITLE
Fix philox randn to follow standard normal distribution

### LINKED ISSUE #91944: [Inductor] philox randn doesn't follow standard normal distribution
### 🐛 Describe the bug

`Philox4_32` `randn` for cpu generates the same value (~1.75) for different `seed`. 
Hence, `Philox4_32` `randn` is (1) not random; (2) not follow standard normal distribution.

```cpp
#include <ATen/core/PhiloxRNGEngine.h>
#include <iostream>

int main() {
  for (int i = 1; i <= 1000; ++i) {
      at::Philox4_32 engine(i);
      std::cout << engine.randn(10) << std::endl; // 1.75668
  }
}
```

I think the root cause of (1) is due to resetting the state every time of `randn` https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/PhiloxRNGEngine.h#L108. Please help provide some context behind this call there.

The root cause of (2) is because `normalize_pair_uniform`  https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/PhiloxRNGEngine.h#L213-L226 is incorrect.

### Additional Context
This issue propagates to TorchInductor `randn` codegen defined with `Philox4_32` `randn` https://github.com/pytorch/pytorch/blob/master/torch/_inductor/codegen/cpp_prefix.h#L33-L36. See comment here https://github.com/pytorch/pytorch/pull/91207#discussion_r1065295810. 

```python
import torch

def foo(x):
    return torch.randn(x.size())

opt_foo = torch.compile(foo)
x = opt_foo(torch.randn(10000))
print(x)
print("mean: ", torch.mean(x)) # 1.7567, but E[X] should be close to 0
print("std: ", torch.std(x)) # 2.3843e-07 (i.e., is not random), but SD[X] should be close to 1

```

### Versions

torch==2.0.0a0+gitb8f35ec

cc @ezyang @gchanan @zou3519 @pbelevich @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by min-jean-cho**
Fixes #91944
Related #91207

### CODE REVIEW

File: aten/src/ATen/core/PhiloxRNGEngine.h [hunk 1]
```diff
@@ -211,17 +210,19 @@ class philox_engine {
...
 
     float mag = std::sqrt(-2.0 * std::log(u1));
 
     detail::FLOAT2 ret;
 
-    ret[0] = mag * std::cos(two_pi);
-    ret[1] = mag * std::sin(two_pi);
+    ret[0] = mag * std::cos(two_pi * u2);
+    ret[1] = mag * std::sin(two_pi * u2);
     return ret;
```
Comment (jgong5) [inline aten/src/ATen/core/PhiloxRNGEngine.h:226]: Only `ret[0]` is used from the caller side, why do we bother to compute two values?

**Review by jgong5 — COMMENTED** [review]
Add UT?

File: aten/src/ATen/core/PhiloxRNGEngine.h (hunk 1 above)
Comment (min-jean-cho) [reply aten/src/ATen/core/PhiloxRNGEngine.h:226]: I think it was because we need two samples from uniform distribution `u1`, `u2` anyways. But, we don't need to bother computing two normal.

**Comment by min-jean-cho** [comment]
> Add UT?

UT for verifying different seed, different random value is simple (I see same seed, same random value is already tested here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/cpu_generator_test.cpp#L251-L255).

But I'm not sure if there's a good way to UT the `randn` values follow normal distribution. One simple way is taking lots of sample and verifying the mean is close to 0 and std is close to 1.

File: aten/src/ATen/core/PhiloxRNGEngine.h [hunk 2]
```diff
@@ -105,15 +105,14 @@ class philox_engine {
     #ifdef __CUDA_ARCH__
     AT_ASSERT(false, "Unsupported invocation of randn on CUDA");
     #endif
-    reset_state(); // Reset state for randn - a little wasteful, but easier to ensure correctness.
     detail::UINT4 counter = counter_;
     detail::UINT2 key = key_;
     detail::UINT4 i = rand(counter, key, n_rounds);
```
Comment (ngimel) [inline aten/src/ATen/core/PhiloxRNGEngine.h:110]: it looks like `rand` doesn't call `incr()` the way operator `()` does, why is that? We are supposed to advance counter for each call.

Comment (min-jean-cho) [reply aten/src/ATen/core/PhiloxRNGEngine.h]: Done, perhaps @voznesenskym could help provide some context why `reset_state()` instead of `incr()` for `randn` were there.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/core/PhiloxRNGEngine.h (hunk 2 above)
Comment (min-jean-cho) [reply aten/src/ATen/core/PhiloxRNGEngine.h]: Hi @voznesenskym, @ngimel ~ bump on this. Is this good to go?

**Review by ngimel — APPROVED** [review]
Yeah lgtm
